docs/adr/reconstruction.md

```diff
@@ -0,0 +1,488 @@
+   - FNs collect samples from LNs.
+   - FNs send bitmap updates to connected FNs.
+
+   <img src="https://github.com/user-attachments/assets/984bedff-8bfc-49df-a35e-be70a05a230a" width="50%">
+
+5. **Sample exchange**
+   - FNs collect samples that are available in the network but do not request those already held locally.
+
+   <img src="https://github.com/user-attachments/assets/1cb1a1a1-7bb1-4f46-afb8-1e2b494e8d7f" width="50%">
+
+6. **Sample relaying and subsequent exchange rounds**
+   - After receiving samples in the first round, FNs continue to send bitmap updates of newly collected samples to other FNs.
+   - FN2 acquires some samples not available on FN1 or FN3, which then request these samples from FN2.
+   - Additional exchange rounds may occur if needed.
+
+   <img src="https://github.com/user-attachments/assets/afa3e44d-c1c7-4163-9ea4-18c4040273f9" width="50%">
+
+7. **New LN joins the network**
+   - Not enough samples are available for each FN to reconstruct the block.
+   - A new LN joins with a fresh set of samples and starts sending them to FN2.
+
+   <img src="https://github.com/user-attachments/assets/2849d1a1-07bd-4284-b1ce-bcaebf713acd" width="50%">
+
+8. **Reconstruction is possible now**
+   - FN2 can now reconstruct the block via erasure decoding.
+
+   <img src="https://github.com/user-attachments/assets/51e81ccb-2107-45d7-b553-f7af1f1cada3" width="50%">
+
+9. **Completed reconstruction notification**
+   - FN2 notifies other FNs that it has finished reconstruction and now holds all samples.
+   - FN1 and FN3 can request any missing samples from FN2. For instance, they might request the sample at (X:1, Y:0).
+
+   <img src="https://github.com/user-attachments/assets/8a0ef334-95fb-4180-aab2-63bb50ebf1da" width="50%">
+
+10. **Reconstruction complete on all FNs**
+- All FNs successfully reconstruct the block.
+
+   <img src="https://github.com/user-attachments/assets/25c5279b-e75d-415a-8c33-5f14c9c9082c" width="50%">
+
+---
+
+### Protocol Diagrams
+
+Below is an outline of the proposed protocols. Full flow diagrams appear at the end of this document.
+
+```
+1. Bitmap Subscription 
+   Client (FN)                            Server (FN)
+      |---- Subscribe to bitmap -------------->|
+      |<---- Initial bitmap -------------------|
+      |<---- Updates  -------------------------|
+      |<---- End updates ----------------------|
+
+2. GetSamples
+   Client (FN)                     Server (FN/LN)
+      |---- Request(bitmap) ----------->|
+      |<---- [Samples + Proof] parts ---|
+```
+
+---
+
+### No Bitmap Subscription From Light Nodes: Why?
+
+Subscribing to bitmaps from Light Nodes would allow more precise deduplication of requested samples, but it also introduces additional complexity and round-trip overhead. Since each Light Node holds relatively few samples, the probability of overlap is low. Instead, the Full Node can send an inverse “have” bitmap in its request to indicate which samples are still needed.
```

Comment:
Does this mean that the FN needs to regularly send the bitmap (or a diff of the bitmap) again to LNs as the FN collects more samples? Or is it not necessary as most LNs only have a few samples anyway?

Reply:
Good question! The short answer is that **updates are not needed.**

Because each Light Node typically holds only a few samples, a single request round-trip to the LN is usually sufficient to retrieve everything it has. If the LN sampled only some of 16 coordinates it can provide none. So the FN can retry later with an increasing backoff. This approach allows the FN to determine whether a particular LN has fully sampled the block and no additional bitmap exchanges are required. Essentially, once the LN reports having all samples (or delivers all it has), the FN can safely move on.

Reply:
How does the FN know if the LN has all the samples though? SamplesRequest struct doesn't provide for this.

Reply:
I think this will result cases where the block cannot be reconstructed even if LNs have enough shares to do so -- for example if 1/4 of the block is withheld, most LN sampling will fail, til reconstruction happens. It's a chicken and egg situation...

Reply:
Yes, that’s correct: if 1/4 of the block is withheld, LNs can’t finish sampling and therefore won’t serve any shares. 99% of LNs will treat the block as unavailable. Although it may seem like a “chicken and egg” issue, the protocol is actually protecting LNs from finalizing incomplete blocks. The block could be recovered if the missing shares eventually surface, but until then, it remains withheld. The assumption here is that protocol’s main goal is to prevent LNs from accepting blocks that aren’t fully available.

Do you think this scenario is a viable attack vector? In given case random 1% of LN will get into thinking that block is available, which is the same as the chance of false positive sampling

Reply:
My point is that in such a scenario, LNs should be sharing their partial samples with FNs, in case FNs can reconstruct the missing parts of the block, and complete the rest of their samples.